Observer tests improvements

4O4 opened the pull request:
Basically all tests are rewritten here. Implementation details should
not be checked because tests are becoming fragile when it comes to any
refactoring.

What instead should be tested is the actual **public behavior** of the
library when using its public interfaces.

While doing all these changes the following assumptions about this codebase
have been made:
 - class fields which are prefixed with underscore (like `observer._onNext`)
 are intended to be private and thus should only be accessed by members of
 the same class or its subclasses
 - `observer.stopped` is a public **readonly** field and should never be
 modified by the outside world, because its role is considered to be
 purely informational

If the above assumptions are not correct, this will need to be discussed
further. These improvements of the tests are necessary to introduce any
changes which include refactoring of the `Observer` in order to fix some
bugs.

4O4 commented:
@bjornbytes do you have some ideas why build fails on LuaJIT2? What are the differences between these runtimes that make this one test fail?

4O4 commented:
Well it looks like a bug in the test framework so I made a quick fix and sent PR for that too: https://github.com/bjornbytes/lust/pull/3
When / if this patch for lust is merged, updating lust version embedded in RxLua repo will resolve problem with that one grumpy test failing on LuaJIT

bjornbytes commented:
Yeah, there's an annoying difference between the different Lua flavors:

```lua
print(#{1, nil, 3}) --> 1 in LuaJIT, 3 in Lua
```

Even though `nil`s are generally fine in observables, it might be better to just avoid them here and use something like `false` instead, to make testing easier.

Though, it seems like the `strict_eq` function would work fine without the length or `tableSize` tests.  For the lust PR I'd kinda prefer just removing the length tests altogether to keep things simpler.

Also, this is a minor thing, but could you remove trailing whitespace in this changeset?  Usually it's possible to configure text editors to do this automatically.

Thanks!!

4O4 commented:
I was actually suspecting the problem was deeper when comparing the extra hash table `{ key = 6 }` I've added as last argument, but yeah you're right - it's that nil's fault. So it seems like this test must have been failing anyway on LuaJIT even before my changes.

I feel like the proper solution here is to fix the test framework instead of adjusting the input data in specs to satisfy the runner. This way tests will be more robust and may actually catch some accidental errors introduced during further refactoring. I've updated the PR on lust project to remove the table size check. 

Can you point out where are the trailing whitespaces you're mentioning? I don't see any signs of them... usually they are noticeable and clutter local diff and / or diff view here on github but I don't see anything strange now.

bjornbytes commented:
Here's what I'm seeing:

![image](https://user-images.githubusercontent.com/784805/66275323-37998300-e83c-11e9-8e47-2f75e5d8315f.png)

4O4 commented:
Whitespaces are fixed I think. It would be best now to first merge #48 and I can then rebase this branch to see if cheks are passing. If you prefer to merge without any particular order I'm fine with that too :)
